interfaces/builtin/desktop_session_control.go

```diff
@@ -0,0 +1,95 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+const desktopSessionControlSummary = `allows setting up user processes for a desktop session`
+
+const desktopSessionBaseDeclarationSlots = `
+  desktop-session-control:
+    allow-installation:
+      slot-snap-type:
+        - app
+        - core
+    deny-auto-connection: true
+`
+
+const desktopSessionControlConnectedPlugAppArmor = `
+# Description: Can control the processes in desktop sessions, launch them, etc.
+
+/var/lib/snapd/desktop/applications/{,**} r,
+
+#include <abstractions/dbus-session-strict>
+#include <abstractions/dbus-strict>
+
+# Supporting session boot fully driven by user session systemd
+# and D-Bus activation
+dbus (send)
+    bus=session
+    path={/,/org/freedesktop/DBus}
+    interface=org.freedesktop.DBus
+    member=UpdateActivationEnvironment
+    peer=(label=unconfined),
+
+dbus (send)
+    bus=session
+    path=/org/freedesktop/systemd1
+    interface=org.freedesktop.DBus.Properties
+    member=Get{,All}
+    peer=(label=unconfined),
+
+dbus (send)
+    bus=session
+    path=/org/freedesktop/systemd1
+    interface=org.freedesktop.systemd1.Manager
+    member={SetEnvironment,UnsetEnvironment,UnsetAndSetEnvironment,ResetFailed,Reload,ListUnitFiles,ListUnitFilesByPatterns,StartUnit,StartTransientUnit}
+    peer=(label=unconfined),
```

Comment:
Just pointing this out so it doesn't get missed in the security review: `StartTransientUnit` allows trivial sandbox escape: the caller provides a new systemd unit as an argument, which can have an arbitrary `ExecStart=` line. The `SetEnvironment` can alter the behaviour of units started in the future (e.g. by setting `LD_PRELOAD`).

This isn't necessarily a reason to reject the PR outright: these permissions could still make a lot of sense for a snap acting as a desktop session of an Ubuntu Core system. We would probably want to lock it down in the base declaration similar to `snapd-control`. Try adapting `snapdControlBaseDeclarationPlugs` for this interface.

Reply:
done, added comments and added a base declaration plug